Introduce new query language (Approach 1) (#220)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

This specification defines the following new parameters:

`presentation_definition`:

> : A string containing a Presentation Definition JSON object. See (#request_presentation_definition) for more details. Exactly one of `vp_query` or `presentation_definition` or `presentation_definition_uri` MUST be present in the Authorization Request.

Reviewer: ```suggestion
: A string containing a Presentation Definition JSON object. See (#request_presentation_definition) for more details. Exactly one of `vp_query`, `presentation_definition`, or `presentation_definition_uri` MUST be present in the Authorization Request.
```

`presentation_definition_uri`:

> : A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. See (#request_presentation_definition_uri) for more details. Exactly one of `vp_query` or `presentation_definition` or `presentation_definition_uri` MUST be present in the Authorization Request.

Reviewer: ```suggestion
: A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. See (#request_presentation_definition_uri) for more details. Exactly one of `vp_query`, `presentation_definition`, or `presentation_definition_uri` MUST be present in the Authorization Request.
```

`vp_query`:

> : A string containing a JSON-encoded VP Query as defined in (#vp_query). This parameter is OPTIONAL. Exactly one of `vp_query` or `presentation_definition` or `presentation_definition_uri` MUST be present in the Authorization Request.

Reviewer 2: ```suggestion
: A string containing a JSON-encoded VP Query as defined in (#vp_query). Exactly one of `vp_query` or `presentation_definition` or `presentation_definition_uri` MUST be present in the Authorization Request.
```

Reviewer: Including @awoie's change
```suggestion
: A string containing a JSON-encoded VP Query as defined in (#vp_query). Exactly one of `vp_query`, `presentation_definition`, `presentation_definition_uri` MUST be present in the Authorization Request.
```

If the Verifier responds with any HTTP error response, the Wallet MUST terminate the process.

# Verifiable Presentation Query Language {#vp_query}

The Verifiable Presentation Query Language (VP Query) is a JSON-encoded query
language that allows the Verifier to request the Wallet to present Verifiable
Credentials that match the query. The query consists of a set of constraints
that the requested Verifiable Credentials must satisfy. The Wallet evaluates the
query against the Verifiable Credentials it holds and returns the Verifiable
Presentations that match the query.

A valid VP Query is defined as a JSON-encoded object with the following
properties:

> `credentials`:

Reviewer 3: i think it is pretty common to use strings even when they are not meant for human consumption. it has also been a convention throughout openid4vc specs: credential_configuration_id and credential_identifier in VCI just being some of the examples. I do not see why deviate from that here and would prefer to keep it a string.

: REQUIRED. An object. Each key in the object represents one
Credential Query as defined in (#credential_query). The key is a string that
uniquely identifies the Credential Query within the VP Query.

`rules`:
: OPTIONAL. An object defining rules that specify which combinations of
credentials MAY be returned by the Wallet, as defined in (#credential_rules). If
omitted, all credentials listed in `credentials` are requested for presentation.

> ## Credential Query {#credential_query}

Reviewer 4: The credential query structure is defined as a polymorphic structure that contains both elements that always have the same content (purpose), elements that are partly format specific (claims) and elements that are fully format specific (everything else).
Since there is only a single element that is not currently format specific (purpose), we should consider making the request fully profile specific to simplify implementations.

Reviewer 3: assuming by profile specific, you meant format specific? I do not understand how making everything format specific would simplify things... I think the feedback we have been getting is actually in the opposite direction of removing "partly format specific" part and use `path` for mdocs too, instead of `namespaces` and `data_elements` claims. the rest of the structure of defining the framework in a format agnostic way, while allowing format specific content for certain claims (like for the algorithms supported) makes sense to me. path vs namespace/data_elements feels pretty unnatural.

Appendix A of [@!OpenID.VCI]. The value of this property MUST be one of the
supported Credential formats as defined in the Wallet's metadata.

Format-specific properties (using the format value as a key):
: OPTIONAL.
Additional properties requested by the Verifier that are specific to the
Credential format, as defined in (#format_specific_properties). If omitted,
no format-specific restrictions are placed on the requested Credential.

`purpose`:
: OPTIONAL. A purpose description array as defined in
(#purpose_description).

`claims`:
: OPTIONAL. An array that specifies the claims that the requested
Verifiable Credential must contain. Each entry in the array corresponds to at
least one claim that MUST be returned by the Wallet to fulfill the credential
request (unless specified otherwise by extensions to this specification). Each
entry MUST be one of the following objects:

- Claim query: An object describing a single claim that is to be returned, as
  specified in (#claim_query).
- An optionality rule: An object containing the following properties:
  - `required` (REQUIRED): defines a number of claims that must be returned from
    the `from` set of claims, as specified next.

>   - `from` (REQUIRED): An array; each entry is either a claim query or another

Author: I'm not sure I fully understand the proposal - could you please give an example?

>   - `from` (REQUIRED): An array; each entry is either a claim query or another

Reviewer 4: Hereby some examples on what it could look like. Option 1 is the current proposal, option 2 and 3 are alternatives. 

```
; option (1) What we have in the current proposal

claims = [+claim]   ; you can ask for multiple "claim", each "claim" MUST be satisfied

claim = claim_query / optionality_rule  ; claim_query OR optionality_rule
optionality_rule = {
    "required": uint,   ; number of claims to be returned from the 'from' set
    "from": claim       ; recursively contains the claim structure
}

claim_query = claim_query_json / claim_query_mdoc   ; claim_query_json OR claim_query_mdoc

claim_query_json = {
    "path": tstr
}

claim_query_mdoc = {
    "namespace": tstr,
    "claim_name": tstr

}

; option (2) For each claim, define the main element, then define alternates for that claim

claims = [+claim]   ; you can ask for multiple "claim", each "claim" MUST be satisfied

claim = {
    "claim_query" = claim_query,            ; claim_query as defined in (1)
    "alternate_claims" = [+ claim_query]    ; Alternate claims, in order of preference
}

; option (3) For each claim, define an array 

claims = [+claim]       ; you can ask for multiple "claim", each "claim" MUST be satisfied
claim = [+claim_query]  ; to satisfy the "claim" array, one element from the array (e.g. one claim_query) must be returned

```

Note, what is the best solution also depends on the outcome of the other discussions. Specifically the ones regarding intent_to_retain, if_present and whether to have the claim_query be polymorphic or to have different versions of the "claims" element, e.g. claims_json and claims_mdoc.

>   - `from` (REQUIRED): An array; each entry is either a claim query or another

Reviewer 4: As discussed yesterday, hereby some examples for these structures. Note that other solutions besides option 2 or option 3 may exist that are more appropriate, especially when we consider the other comments on this PR.

```
Example 3: given_name, age_over_18 OR birth_date OR age_birth_year

option(1)

claims = [
    { "path": ["given_name"] },
    {
        "required": 1,
        "from": [
            { "path": ["age_over_18"] },
            { "path": ["birth_date"] },
            { "path": ["age_birth_year"] }
        ] 
    }
]

option(2)

claims = [ 
    {
        "claim_query": { "path": ["given_name"] }   
    },
    {
        "claim_query": { "path": ["age_over_18"] },
        "alternate_claims": [
            { "path": ["birth_date"] },
            { "path": ["age_birth_year"] }
        ]  
    },  
]

option(3)

claims = [ 
    [ { "path": ["given_name"] } ]
    [             
        { "path": ["age_over_18"] },
        { "path": ["birth_date"] },
        { "path": ["age_birth_year"] }
    ]
]

Example 4: given_name, zip_code OR (city AND state)

option(1)

claims = [
    { "path": ["given_name"] },
    {
        "required": 2,
        "from": [
            { "path": ["zip_code"] },
            {
                "required": 1,
                "from": [
                    { "path": ["city"] },
                    { "path": ["state"] },
                ]
            }
        ] 
    }
]

option(2)

claims = [ 
    {
        "claim_query": { "path": ["given_name"] }   
    },
    {
        "claim_query": { "path": ["zip_code"] },
        "alternate_claims": [
            { "path": ["city"] },
        ]  
    },  
    {
        "claim_query": { "path": ["zip_code"] },
        "alternate_claims": [
            { "path": ["state"] },
        ]  
    }
]

option(3)

claims = [ 
    [ { "path": ["given_name"] } ]
    [             
        { "path": ["zip_code"] },
        { "path": ["city"] },
    ],
    [             
        { "path": ["zip_code"] },
        { "path": ["state"] },
    ]
]
```

>   - `from` (REQUIRED): An array; each entry is either a claim query or another

Reviewer 5: Does `,` in the above mean AND/required? I think the description text should be more specific and informative.
Also, does the `Example 4` mean this `given_name AND (zip_code OR (city AND state))`
seems like in option 3 `zip_code OR city` and `zip_code OR state` could result in zip_code and zip_code. Resulting in not returning city or state.

Can we redefine option 3 by swapping the meaning of things in the 1st and 2nd level `[ ]`?
E.g. given_name AND (zip_code OR (city AND state))
```
redefining option 3
claims = [ 
    [ 
        { "path": ["given_name"] },
       { "path": ["zip_code"] },
    ]
    [             
       { "path": ["given_name"] },
       { "path": ["state"] },
       { "path": ["city"] },
    ],
]
```

This would make it simpler as entries in the root level `[ ]` would mean OR and the 2nd level will mean MUST.
In the above it translates to `(given_name and zipcode) OR (given_name and zip_code and city)`

Similarly for `Example 3: given_name and (age_over_18 OR birth_date OR age_birth_year)`
```
redefining option 3
claims = [
  [ 
        { "path": ["given_name"] },
       { "path": ["age_over_18"] },
  ],
 [ 
        { "path": ["given_name"] },
       { "path": ["birth_date"] },
  ],
  [ 
        { "path": ["given_name"] },
       { "path": ["age_birth_year"] },
  ],
```
This also corresponds to all the potential credentials User can consent to send. So it won't result in cases where `zip_code and zip_code` are sent.
I recognize that this could cause the query to be longer than other methods in few cases.

>   - `from` (REQUIRED): An array; each entry is either a claim query or another

Reviewer 5: I think it would be more clear to use a notation (req) instead of `and` 
as in boolean "zip_code" and "zip_code" translates to just 1 "zip_code"

https://grace.bluegrass.kctcs.edu/~kdunn0001/files/Simplification/4_Simplification_print.html

Where as in (req)zip_code (req)zip_code it is sent twice?

More than two nested levels of optionality rules MUST NOT be used by a Verifier
or allowed by a Wallet.

## Format-specific Properties {#format_specific_properties}

The following format-specific properties are defined:

### Format `vc+sd-jwt` {#format_vc_sd_jwt}

`vct_values`:
: OPTIONAL. An array of strings that specifies allowed values for
the type of the requested Verifiable Credential. All elements in the array MUST
be valid type identifiers as defined in [@!I-D.ietf-oauth-sd-jwt-vc]. The Wallet
MAY return credentials that inherit from any of the specified types, following
the inheritance logic defined in [@!I-D.ietf-oauth-sd-jwt-vc].

`credential_signing_alg_values`:
: OPTIONAL. An array of strings that specifies
the allowed algorithms for signing the SD-JWT (i.e., only the issuer-signed part).

`kbjwt_signing_alg_values`:
: OPTIONAL. An array of strings that specifies the
allowed algorithms for signing the Key-Binding JWT (KB-JWT) (i.e., the part signed
by the holder).

`sd_alg_values`:
: OPTIONAL. An array of strings that specifies the allowed hash algorithms for
digests in the SD-JWT and KB-JWT.

### Format `mso_mdoc` {#format_mso_mdoc}

TBD

> ### Format `jwt_vp*`

Reviewer 3: ```suggestion
### Format `jwt_vp`
```

`namespace`:
: REQUIRED. A string that specifies the namespace of the claim
within the Verifiable Credential, e.g., `org.iso.18013.5.1`.

`claim_name`:
: REQUIRED. A string that specifies the name of the claim within the provided namespace
in the Verifiable Credential, e.g., `first_name`.

### Claims Path Pointer {#claims_path_pointer}

A claims path pointer is a pointer into the JSON structure of the Verifiable
Credential, identifying one or more claims. A claims path pointer MUST be a
non-empty array of strings and non-negative integers. A `*` string value indicates
that all elements of the currently selected array(s) are to be selected; any other
string indicates that the respective key is to be selected; and a non-negative
integer indicates that the respective index in an array is to be selected. The path
is formed as follows:

  - Start with an empty array.
  - To address a particular claim within an object, append the key (claim name)
    to the array.
  - To address an element within an array, append the index to the array (as a
    non-negative, 0-based integer).
  - To address all elements within an array, append the string `*` to the array.

> Verifiers MUST NOT point to the same claim more than once in a single query.
> Wallets SHOULD ignore such duplicate claim queries.

Reviewer: I think it very likely that verifiers will accidentally point to the same claim more than once, likely when mixing claim indexes with claim names, and certainly when mixing the wildcard with either or both claim indexes and/or claim names.

It might be viable to forbid use of the wildcard in combination with integers and/or strings, but that is not sufficient to cover all scenarios.

```suggestion
Verifiers SHOULD NOT point to the same claim more than once in a single query,
e.g., by combining the wildcard `*` with one or more claim names or indexes,
or by using claim names *and* indexes.
Wallets SHOULD return each claim once, regardless of duplicate claim queries.
```

Reviewer 6: Hi @TallTed I've hidden this comment because as I replied to a previous comment a week or two ago we don't appear to have a contribution agreement in place for you so can't use your contributions. If you'd like you contribute you can complete the contribution agreement here: https://openid.net/intellectual-property/openid-foundation-contribution-agreements/

If you are already covered by a contribution agreement, please let me know the details so we can locate it. Thank you!

#### Example

The following shows a non-normative, simplified example of a Credential:

```json
{
  "name": "Arthur Dent",
  "address": {
    "street_address": "42 Market Street",

>     "city": "Milliways",

Reviewer 7: ```suggestion
    "locality": "Milliways",
```

       selected array(s). If any of the currently selected element(s) is not an
       array, abort processing and return an error.
    3. If the query component is a non-negative integer, select the element at
       the respective index in the currently selected array(s). If any of the
       currently selected element(s) is not an array, abort processing and
       return an error. If the index does not exist in a selected array, remove
       that array from the selection.
  3. If the set of elements currently selected is empty, abort processing and
     return an error.

The result of the processing is the set of elements which is requested for
presentation.

## Purpose Description {#purpose_description}

A purpose description array is an object that describes the purpose for which
the Verifier is requesting the Verifiable Credential. It can contain multiple
objects, each describing the purpose in a specific language. Each object
contains the following properties:

`lang:`
: REQUIRED. A string that specifies the language of the purpose
description. The value of this property MUST be a valid language tag as defined

> in [@!BCP47].

Reviewer 7: I suggest using the identifier `locale` as is done in OpenID Connect.  The description there is
> End-User's locale, represented as a BCP47 [RFC5646] language tag. This is typically an ISO 639 Alpha-2 [ISO639] language code in lowercase and an ISO 3166-1 Alpha-2 [ISO3166‑1] country code in uppercase, separated by a dash. For example, en-US or fr-CA. As a compatibility note, some implementations have used an underscore as the separator rather than a dash, for example, en_US

`text`:
: REQUIRED. A string that contains the purpose description in the
language specified by the `lang` property.

## Credential Rules {#credential_rules}

> Just as for claims, the Verifier can define rules that specify which

Reviewer 8: We are using the PE submission_requirements feature. I see the intent here is to replace `submission_requirements` with `rules`. The features that we in HR use cases that are missing from rules are

1. ability to require `all` credentials of a particular credential type (eg. all transcripts) vs one of a particular type (eg mDL). From my understanding both would expressed as `{ "required": 1, "from": [{"ref": "transcript"}]}` and `{ "required": 1, "from": [{"ref": "mDL"}]}`
2. in HR use cases the user has more flexibility in what to share. therefore the ability to also require at least one is useful `{"min": 1, "from": [{"ref": "email"}]}`.
3. explaining the purpose of  a `rule` (e.g. the RP requires a mDL or national id card to get the applicants name)
4. is there a maximum depth to rules?

Reviewer 8: > ability to require all credentials of a particular credential type (eg. all transcripts) vs one of a particular type (eg mDL). From my understanding both would expressed as { "required": 1, "from": [{"ref": "transcript"}]} and { "required": 1, "from": [{"ref": "mDL"}]}
in HR use cases the user has more flexibility in what to share. therefore the ability to also require at least one is useful {"min": 1, "from": [{"ref": "email"}]}.

As per a simiilar suggestion above `required` should be `all`, `optional`, `one`, `at_least_one` with the default either being `all` or `optional`. An alternative if a number is required for certain use cases would be to replace `one` with `count` which would then have an additional property with the expected `count`

Author: The `at_least_one` case is not currently captured, but we should consider introducing it and therefore moving back to keywords instead of numbers. Do you have a concrete use case in mind for `at_least_one`?

Author: How would a purpose for the rule be exposed to the user?

Reviewer 8: @danielfett the current use case we have is in the application for a job. In the recruitment use case the holder can share whatever information they feel is relevant. For example, a person may have had studied and achieved a number of diplomas, but one or more maybe irrelevant and/or perhaps detrimental to them being successful in the application. The recruiter therefore will request at least one degree but leave it up to the holder to select the relevant one or two. 

Again the recruitment use case means that certain types of credentials are needed for a purpose. We have a UI in the reference wallet in the HR use case that takes the user through a wizard-like process to meet each rule requirement. A purpose for that rule is required for this UX to make sense

combinations of credentials MAY be returned by the Wallet. The rules are defined
in the top-level property `rules`. The values is a credential rule object,
defined as having the following properties:

- `required`: An integer specifying the number of credentials that MUST be
  returned from the `from` set of credentials.
- `from`: An array of objects, each being either
  - a nested credential rule, or
  - a credential query reference, i.e., an object with a single key `ref` that
    references a credential query defined in the `credentials` object.

Within `from`, the credentials are listed in order of preference, with the first
entry being the most preferred.

## Examples {#vp_query_examples}

The following is a non-normative example of a VP Query that requests a Verifiable
Credential of the format `vc+sd-jwt` with a type value of
`https://credentials.example.com/identity_credential` and the claims `last_name`,
`first_name`, and `address.street_address`:

<{{examples/query_lang/simple.json}}

Additional examples can be found in (#vp_query_examples).

## Additional Request Properties {#additional_request_properties}

> `intend_to_retain`: TBD

Reviewer 9: ```suggestion
`intent_to_retain`: TBD
```

If the Verifier responds with any HTTP error response, the Wallet MUST terminate the process.

> # Verifiable Presentation Query Language {#vp_query}

Reviewer 10: I would like to see this as a separate specification. It can be fully tested outside of OID4VP flows.

Reviewer 3: i don't think the intention of this new query language has been to be protocol independent, so not sure what would be the benefit of "fully testing outside of OID4VP flows" unless I am misinterpreting it..

Reviewer 10: My thought is that a software implementation of a query language is distinct from an implementation for OID4VP. It could have its own conformance suite and open source tooling, and potentially, utility outside of this specification.

If those broader ambitions are shared amongst the group then breaking it out could be the right path.

uniquely identifies the Credential Query within the VP Query.

`rules`:
: OPTIONAL. An object defining rules that specify which combinations of
credentials MAY be returned by the Wallet, as defined in (#credential_rules). If
omitted, all credentials listed in `credentials` are requested for presentation.

## Credential Query {#credential_query}

A Credential Query is an object representing a request for one specific
credential. It contains the following properties:

`format`:
: REQUIRED. A string that specifies the format of the requested
Verifiable Credential. Valid Credential format identifier values are defined in
Appendix A of [@!OpenID.VCI]. The value of this property MUST be one of the
supported Credential formats as defined in the Wallet's metadata.

Format-specific properties (using the format value as a key):
: OPTIONAL.
Additional properties requested by the Verifier that are specific to the
Credential format, as defined in (#format_specific_properties). If omitted,
no format-specific restrictions are placed on the requested Credential.

`purpose`:
: OPTIONAL. A purpose description array as defined in
(#purpose_description).

> `claims`:

Reviewer 2: I'm wondering whether it is a good idea to have `claims` being a polymorphic object with a format-dependent object type, or if it would be better to have everything nested inside a format identifier object, e.g., `sd-jwt-vc: { claims ... }`?

Author: I would not put it into the format-specific part, as `claims` applies to either all JSON-based structures (which can be multiple formats) or all ISO-based structures (mdoc). We could rename to `claims` and `claims_iso` maybe.

: OPTIONAL. An array that specifies the claims that the requested
Verifiable Credential must contain. Each entry in the array corresponds to at
least one claim that MUST be returned by the Wallet to fulfill the credential
request (unless specified otherwise by extensions to this specification). Each
entry MUST be one of the following objects:

- Claim query: An object describing a single claim that is to be returned, as
  specified in (#claim_query).
- An optionality rule: An object containing the following properties:
  - `required` (REQUIRED): defines a number of claims that must be returned from
    the `from` set of claims, as specified next.
  - `from` (REQUIRED): An array; each entry is either a claim query or another
    optionality rule. The entries are listed in order of preference, with the
    first entry being the most preferred.

> More than two nested levels of optionality rules MUST NOT be used by a Verifier

Reviewer 10: if credentials have >2 layers of nesting what's the solution?

Author: This works for any depth of nesting; the optionality rules are just capped to enable easier processing, but this doesn't reduce expressiveness of the language.

Reviewer 4: Does "two nested levels of optionality rules" mean that there is a total of 2 or 3 instances of "required" in a single "claim" element?

or allowed by a Wallet.

## Format-specific Properties {#format_specific_properties}

The following format-specific properties are defined:

> ### Format `vc+sd-jwt` {#format_vc+sd-jwt}

Reviewer 10: are we identifying formats by media type?
this is an unregistered media type

it may be better to have unique format identifiers that point to versions of specifications, defined by this specification.

Reviewer 7: Many of the JWT claim names used in the examples are wrong.  For instance, `first_name` is used, whereas the correct claim name is `given_name`.  I suggested changes.

the type of the requested Verifiable Credential. All elements in the array MUST
be valid type identifiers as defined in [@!I-D.ietf-oauth-sd-jwt-vc]. The Wallet
MAY return credentials that inherit from any of the specified types, following
the inheritance logic defined in [@!I-D.ietf-oauth-sd-jwt-vc].

`credential_signing_alg_values`:
: OPTIONAL. An array of strings that specifies
the allowed algorithms for signing the SD-JWT (i.e., only the issuer-signed part).

`kbjwt_signing_alg_values`:
: OPTIONAL. An array of strings that specifies the
allowed algorithms for signing the Key-Binding JWT (KB-JWT) (i.e., the part signed
by the holder).

`sd_alg_values`:
: OPTIONAL. An array of strings that specifies the allowed hash algorithms for
digests in the SD-JWT and KB-JWT.

### Format `mso_mdoc` {#format_mso_mdoc}

TBD

### Format `jwt_vp*`

TBD

## Claim Query {#claim_query}

A claim query is an object that describes one or more claims that are to be
returned. Depending on the format of the requested Verifiable Credential, the
claim query can contain different properties.

### JSON-based Claims Structures {#json_based_claims}

> For credentials containing JSON-based claims structures, the following

Reviewer 2: We should clearly define what credential formats use this section rather than saying credentials using JSON-based claims structures.

Reviewer 3: i would prefer using path for mdocs too to solve this weird deviation between the formats for this top-level parameter layer

properties are defined:

`path`:
: REQUIRED. A claims path pointer that specifies the path to the claim
within the Verifiable Credential, as defined in (#claims_path_pointer).

### ISO 18013-5 Credentials {#iso_18013-5}

For ISO 18013-5 Credentials, the following properties are defined:

`namespace`:
: REQUIRED. A string that specifies the namespace of the claim
within the Verifiable Credential, e.g., `org.iso.18013.5.1`.

`claim_name`:

> : REQUIRED. A string that specifies the name of the claim within
> the Verifiable Credential, e.g., `first_name`.

Reviewer 2: ```suggestion
: REQUIRED. A string that specifies the name of the claim within the provided namespace
in the Verifiable Credential, e.g., `first_name`.
```

### Claims Path Pointer {#claims_path_pointer}

A claims path pointer is a pointer into the JSON structure of the Verifiable
Credential, identifying one or more claims. A claims path pointer MUST be a

> non-empty array of strings and non-negative integers. A string indicates that
> the respective key is to be selected, a `*` value indicates that all elements of
> the currently selected array(s) are to be selected, and a non-negative integer
> indicates that the respective index in an array is to be selected. The path is
> formed as follows:

Reviewer: ```suggestion
non-empty array of strings and non-negative integers. A `*` string value indicates
that all elements of the currently selected array(s) are to be selected; any other
string indicates that the respective key is to be selected; and a non-negative
integer indicates that the respective index in an array is to be selected. The path
is formed as follows:
```

  "nationalities": ["British", "Betelgeusian"]
}
```

The following shows examples of claims path queries and the respective selected
claims:

- `["name"]`: The claim `name` with the value `Arthur Dent` is selected.
- `["address"]`: The claim `address` with its sub-claims as the value is
  selected.
- `["address", "street_address"]`: The claim `street_address` with the value `42
  Market Street` is selected.
- `["degrees", "*", "type"]`: All `type` claims in the `degrees` array are
  selected.
- `["nationalities", 1]`: The second nationality is selected.
- `[]` (empty array): The entire Credential is selected.

#### Processing

In detail, the array is processed by the Wallet from left to right as follows:

 1. Select the root element of the credential, i.e., the top-level JSON object.
 2. Process the query components from left to right:
    1. If the query component is a string, select the element in the respective
       key in the currently selected element(s). If any of the currently
       selected element(s) is not an object, abort processing and return an
       error. If the key does not exist in any element currently selected,
       remove that element from the selection.

>     2. If the query component is `*`, select all elements of the currently

Reviewer: It seems that processing the wildcard should precede any specific key string or integer index, none of which should be processed following the wildcard, as they would then be pointed to twice in the same query.

       array, abort processing and return an error.
    3. If the query component is a non-negative integer, select the element at
       the respective index in the currently selected array(s). If any of the
       currently selected element(s) is not an array, abort processing and
       return an error. If the index does not exist in a selected array, remove
       that array from the selection.
  3. If the set of elements currently selected is empty, abort processing and
     return an error.

The result of the processing is the set of elements which is requested for
presentation.

## Purpose Description {#purpose_description}

A purpose description array is an object that describes the purpose for which
the Verifier is requesting the Verifiable Credential. It can contain multiple
objects, each describing the purpose in a specific language. Each object
contains the following properties:

`lang:`
: REQUIRED. A string that specifies the language of the purpose
description. The value of this property MUST be a valid language tag as defined
in [@!BCP47].

`text`:
: REQUIRED. A string that contains the purpose description in the

> language specified by the `lang` property.

Reviewer 2: IMO, free text is not a good idea for a lot of reasons, e.g., disambiguation between different translations of the the same reason and actually providing different reasons -> prone to errors; privacy; i18n/l18n which is hard to scale with this approach. A more appropriate approach is to define reason codes. There was a lot of work in other standards orgs on that, e.g., Kantara.

Author: Could you please propose a change introducing reason codes so that we can discuss on the calls?

Reviewer 4: In order to provide reason codes that can help users guide in how to respond in a meaningful way, we need to find discuss the proper places where to define such reason codes. These reason codes are likely to be use case and/or document type specific. Therefore one approach could be to define reason codes in such a way that they can be scoped/namespaced according to the entity defining them. 
Separate to that we should also decide whether we can define or reference a set of such codes that could apply more generically.

Author: I must admit that I lack the imagination how a set of codes could ever be complete. "Complete" here meaning in particular that I can use them to refer to all possible reasons for obtaining data through a wallet. Just as a few examples:

- "Required for account opening according to [law XYZ]" — we cannot enumerate all possible laws; also many laws and their applicability will be dependent on the region.
- "Age verification according to [regulatory body XYZ]"
- "Required only for obtaining [product  XYZ]"
- "Alternative when not in possession of [credential XYZ]"

Reviewer 4: Agreed that it's difficult / unlikely to have reasons codes that are both very specific and can address every possible situation. However, allowing free text is not a solution to achieve that as it introduces privacy and security problems that now need to be addressed. it's  a lot more difficult for the wallet to meaningfully use a free form text field to help guide the user compared to pre-defined reason codes.

Reviewer 11: In my experience, providing a meaningful purpose to a user requires customization of the respective text to the particular RP and process, even the specific transaction in some cases. Yes, free text introduces security risks, but those are not unsolvable, developers cope with them every day when building applications with dynamic input. The multi language aspect is a true challenge, it would most likely require the RP sending the purpose string in multiple languages.  
Pre-defined reasons code on the other hand posse their own challenges: 
1) There needs to be registries for those codes. Which parties operate these registries?
2) What does a RP need to do in order to be able to send the first presentation request if this RP wants to implement a use case no one has implemented/standardized before? I think requiring pre-defined reason codes would create a huge administrative burden to RPs and delay deployment of new integrations in an inacceptable manner. 
3) How are those code distributed along with the respective text? What is the update policy? 
4) In the simplest case, there is no link possible to an actual transaction performed by a RP.

Note: The `nonce` and `aud` are set to the `nonce` of the request and the Client Identifier of the Verifier, respectively, in the same way as for the Verifier, Verifiable Presentations to prevent replay.

# Examples for VP Queries {#vp_query_examples}

The following is a non-normative example of a VP Query that requests a Verifiable
Credential in the format `mso_mdoc` with the claims `vehicle_holder` and
`first_name`:

<{{examples/query_lang/simple_mdoc.json}}

The following is a non-normative example of a VP Query that shows the usage of
the `purpose` property, and requests the claim `last_name` and

> - either the claim `zip_code`, or
> - the claims `city` and `state`:

Reviewer 7: ```suggestion
- either the claim `postal_code`, or
- the claims `locality` and `region`:
```
